### PR TITLE
Track the number of tokens seen to metrics

### LINKED ISSUE #27027: Count of tokens seen during training in Trainer
### Feature request

The `Trainer` API should track and log the number of tokens seen during training.

While it sometimes could (maybe?) be possible to back out the number of tokens seen from the FLOS, or by iterating over the whole dataset, it would make a lot of sense for the Trainer API to track the number of tokens seen (and it shouldn't be necessary to completely iterate over a model's training loop just to compute the count of tokens, which is the only current implementation of any token-related metric in Trainer, [`Trainer.num_tokens()`](https://github.com/huggingface/transformers/blob/acc394c4f5e1283c19783581790b3dc3105a3697/src/transformers/trainer.py#L1180)).

This can't currently be implemented in a CallBack, because callbacks don't have access to the training data (only the trainer state).

### Motivation

Number of tokens seen is an essential metric tracked in nearly every LLM training run. It is widely considered one of the fundamental drivers of model quality (tokens seen during training is reported for nearly every major LLM release). It seems that any language model developer using Hugging Face would like to know this metric for their training runs -- it maybe even more important and useful than the FLOS, and perhaps as important as the number of gradient steps. 

In any case, it's an extremely useful number to have, and it must be tracked during training as the model consumes examples.

### Your contribution

I'm willing to contribute this but would like some guidance on the overall design first.

In particular, here's what I think a reasonable implementation would include:

- Add a `global_tokens_seen` or similar to the `TrainerState`. This would add only a single integer value to the `TrainerState`.
- Increment this during `Trainer._inner_training_loop()`
- Probably add this information to the logging outputs

What do the folks at HF think about that?

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR adds `num_tokens_seen` to the `TrainerState` allowing users to know how many tokens were passed in an individual batch. Uses `gather` to ensure that in DDP this can be known as well.

Fixes https://github.com/huggingface/transformers/issues/27027

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@pacman100 @amyeroberts

### CODE REVIEW

**Comment by muellerzr** [comment]
@amyeroberts question: these failures are all because they cannot find the `main_input_name` in the batch. Does this mean some of the models are wrong?

For instance, one of the failures in the examples script is from `SpeechEncoderDecoderModel`, which states `main_input_name` is `"inputs"`, however when running the script it's using `"input_values":

```python
self = {'input_values': tensor([[-0.0177, -0.0188, -0.0202,  ..., -0.0032, -0.0068, -0.0039],
        [-0.0196, -0.0556, -0.0...-100, -100, -100, -100, -100, -100, -100, -100, -100,
         -100, -100, -100, -100, -100, -100, -100, -100, -100]])}
```

File: src/transformers/trainer_callback.py [hunk 1]
```diff
@@ -59,6 +59,8 @@ class TrainerState:
             Run an evaluation every X steps.
         save_steps (`int`, *optional*, defaults to 500):
             Save checkpoint every X updates steps.
+        num_tokens_seen (`int`, *optional*, defaults to 0):
+            The number of tokens seen during training (number of input tokens, not the number of prediction tokens).
```
Comment (amyeroberts) [inline src/transformers/trainer_callback.py:63]: We can disambiguate this by calling this `num_input_tokens_seen`

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1837,6 +1837,8 @@ def _inner_training_loop(
             step = -1
             for step, inputs in enumerate(epoch_iterator):
                 total_batched_samples += 1
+                main_input_name = getattr(self.model, "main_input_name", "input_ids")
+                self.state.num_tokens_seen += self.accelerator.gather(inputs[main_input_name]).numel()
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1841]: Is this something we want to have by default? Firstly because of the additional computation we're adding but also the assumption that `inputs[main_input_name]` represents tokens.

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this!

Main comment is about whether this should be running by default

**Comment by amyeroberts** [comment]
@muellerzr I think we don't have to worry about that if it's behind a flag in the training args. This way - if someone want to measure it then they have to make sure that `main_input_name` is properly set, but it won't break trainer for models which are currently compatible but don't have this correctly set.

**Review by amyeroberts — APPROVED** [review]
Thanks for adding and iterating! 

Just some nits on the implementation

**Comment by muellerzr** [comment]
@amyeroberts failing tests seem to be unrelated

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1838,6 +1838,17 @@ def _inner_training_loop(
...
             for step, inputs in enumerate(epoch_iterator):
                 total_batched_samples += 1
+
+                if self.args.include_num_input_tokens_seen:
+                    main_input_name = getattr(self.model, "main_input_name", "input_ids")
+                    if main_input_name not in inputs:
+                        logger.warning(
+                            "Tried to track the number of tokens seen, however could the current model is "
+                            "not configured properly to know what item is the input. To fix this, add "
+                            "a `main_input_name` attribute to the model class you are using."
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1848]:
```suggestion
                            "Tried to track the number of tokens seen, however the current model is "
                            "not configured properly to know what item is the input. To fix this, add "
                            "a `main_input_name` attribute to the model class you are using."
```

**Review by amyeroberts — APPROVED** [review]
Thanks!
